src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs

```diff
@@ -1837,21 +1838,32 @@ private IDeclarationPatternOperation CreateBoundDeclarationPatternOperation(Boun
                 variable = ((BoundDiscardExpression)boundDeclarationPattern.VariableAccess).ExpressionSymbol;
             }
 
+            ITypeSymbol inputType = boundDeclarationPattern.InputType;
+            bool acceptsNull = boundDeclarationPattern.IsVar;
+            ITypeSymbol matchedType = acceptsNull ? null : boundDeclarationPattern.DeclaredType.Type;
             SyntaxNode syntax = boundDeclarationPattern.Syntax;
-            ITypeSymbol type = null;
-            Optional<object> constantValue = default(Optional<object>);
             bool isImplicit = boundDeclarationPattern.WasCompilerGenerated;
-            return new DeclarationPatternOperation(variable, _semanticModel, syntax, type, constantValue, isImplicit);
+            return new DeclarationPatternOperation(inputType, matchedType, variable, acceptsNull, _semanticModel, syntax, isImplicit);
         }
 
         private IRecursivePatternOperation CreateBoundRecursivePatternOperation(BoundRecursivePattern boundRecursivePattern)
         {
+            ITypeSymbol matchedType = boundRecursivePattern.DeclaredType?.Type ?? boundRecursivePattern.InputType;
             ISymbol variable = boundRecursivePattern.Variable;
+            ISymbol deconstructSymbol = boundRecursivePattern.DeconstructMethod;
+            if (variable == null && boundRecursivePattern.VariableAccess?.Kind == BoundKind.DiscardExpression)
+            {
+                variable = ((BoundDiscardExpression)boundRecursivePattern.VariableAccess).ExpressionSymbol;
+            }
+
+            TypeSymbol inputType = boundRecursivePattern.InputType;
             SyntaxNode syntax = boundRecursivePattern.Syntax;
-            ITypeSymbol type = null;
-            Optional<object> constantValue = default(Optional<object>);
+            ImmutableArray<IPatternOperation> deconstructionSubpatterns =
+                boundRecursivePattern.Deconstruction.SelectAsArray(p => (IPatternOperation)Create(p.Pattern));
```

Comment:
Why can't we do these lazily? Is that just something you're planning on adding laterz/

Reply:
They are now done lazily